src/werkzeug/formparser.py

```diff
@@ -423,156 +411,322 @@ def validate_boundary(self, boundary):
-            # character we have to flush it, otherwise we will chop of
-            # certain values.
-            if buf not in (b"", b"\r", b"\n", b"\r\n"):
-                yield _cont, buf
+        Always obeys the grammar
+        parts = ( begin_form cont* end |
+                  begin_file cont* end )*
+        """
 
-            yield _end, None
+        line_splitter = self.LineSplitter(self.buffer_size if cap_at_buffer else None)
+        line_parser = self.LineParser(self, boundary)
+        while True:
+            buf = file.read(self.buffer_size)
+            lines = line_splitter.feed(buf)
+            parts = line_parser.feed(lines)
+            for part in parts:
+                yield part
+            if buf == b"":
+                break
 
     def parse_parts(self, file, boundary, content_length):
         """Generate ``('file', (name, val))`` and
         ``('form', (name, val))`` parts.
         """
-        in_memory = 0
-
-        for ellt, ell in self.parse_lines(file, boundary, content_length):
-            if ellt == _begin_file:
-                headers, name, filename = ell
-                is_file = True
-                guard_memory = False
-                filename, container = self.start_file_streaming(
-                    filename, headers, content_length
-                )
-                _write = container.write
-
-            elif ellt == _begin_form:
-                headers, name = ell
-                is_file = False
-                container = []
-                _write = container.append
-                guard_memory = self.max_form_memory_size is not None
-
-            elif ellt == _cont:
-                _write(ell)
-                # if we write into memory and there is a memory size limit we
-                # count the number of bytes in memory and raise an exception if
-                # there is too much data in memory.
-                if guard_memory:
-                    in_memory += len(ell)
-                    if in_memory > self.max_form_memory_size:
-                        self.in_memory_threshold_reached(in_memory)
-
-            elif ellt == _end:
-                if is_file:
-                    container.seek(0)
-                    yield (
-                        "file",
-                        (name, FileStorage(container, filename, name, headers=headers)),
-                    )
-                else:
-                    part_charset = self.get_part_charset(headers)
-                    yield (
-                        "form",
-                        (name, b"".join(container).decode(part_charset, self.errors)),
-                    )
+        line_splitter = self.LineSplitter()
+        line_parser = self.LineParser(self, boundary)
+        part_parser = self.PartParser(self)
+        file_writer = self.FileWriter(self, content_length)
+        while True:
+            buf = file.read(self.buffer_size)
```

Comment:
Here is why I recommend generators above. In an async context, it will eventually make sense to yield each event as soon as possible--which means not necessarily waiting until the whole buffer (up to `self.buffer_size`) is consumed. 

So all of these `feed` functions can accept just a single element (instead of multiple ones). They should also avoid resetting their state between calls. In this way, the top-level API here can look like this:

```
while True:
  buf = file.read(self.buffer_size)
  if not buf:  # might be a better way to check this--does file have a "closed" member?
    break
  for line in line_splitter.feed(buf):
    for part in line_parser.feed(line):
      for event in part_parser.feed(part):
        for fevent in file_writer.feed(event):
          yield fevent
```